backend/monoprocess/monoprocess.go

```diff
@@ -0,0 +1,175 @@
+// uses channels to notify the worker every time there is a new task ready to be
+// worked on. Note that only one worker will be notified.
+// IMPORTANT: Only use this backend if the backend and worker are running in the
+// same process.
+func NewMonoprocessBackend(b backend.Backend, signalBufferSize int, signalTimeout time.Duration) *monoprocessBackend {
+	if signalTimeout <= 0 {
+		signalTimeout = time.Second // default
+	}
+	mb := &monoprocessBackend{
+		Backend:        b,
+		workflowSignal: make(chan struct{}, signalBufferSize),
+		activitySignal: make(chan struct{}, signalBufferSize),
+		signalTimeout:  signalTimeout,
+		logger:         b.Logger(),
+	}
+	return mb
+}
+
+func (b *monoprocessBackend) GetWorkflowTask(ctx context.Context) (*task.Workflow, error) {
+	if w, err := b.Backend.GetWorkflowTask(ctx); w != nil || err != nil {
+		return w, err
+	}
+	b.logger.DebugContext(ctx, "worker waiting for workflow task signal")
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-b.workflowSignal:
+		b.logger.DebugContext(ctx, "worker got a workflow task signal")
+		return b.GetWorkflowTask(ctx)
+	}
+}
+
+func (b *monoprocessBackend) GetActivityTask(ctx context.Context) (*task.Activity, error) {
+	if a, err := b.Backend.GetActivityTask(ctx); a != nil || err != nil {
+		return a, err
+	}
+	b.logger.DebugContext(ctx, "worker waiting for activity task signal")
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-b.activitySignal:
+		b.logger.DebugContext(ctx, "worker got an activity task signal")
+		return b.GetActivityTask(ctx)
+	}
+}
+
+func (b *monoprocessBackend) CreateWorkflowInstance(ctx context.Context, instance *workflow.Instance, event *history.Event) error {
+	if err := b.Backend.CreateWorkflowInstance(ctx, instance, event); err != nil {
+		return err
+	}
+	b.notifyWorkflowWorker(ctx)
+	return nil
+}
+
+func (b *monoprocessBackend) CompleteWorkflowTask(
+	ctx context.Context,
+	task *task.Workflow,
+	instance *workflow.Instance,
+	state core.WorkflowInstanceState,
+	executedEvents, activityEvents, timerEvents []*history.Event,
+	workflowEvents []history.WorkflowEvent,
+) error {
+	if err := b.Backend.CompleteWorkflowTask(ctx, task, instance, state, executedEvents, activityEvents, timerEvents, workflowEvents); err != nil {
+		return err
+	}
+	for _, e := range activityEvents {
+		if e.Type != history.EventType_ActivityScheduled {
+			continue
+		}
+		if !b.notifyActivityWorker(ctx) {
+			break // no reason to notify more, queue is full
+		}
+	}
+	for _, e := range timerEvents {
+		attr, ok := e.Attributes.(*history.TimerFiredAttributes)
+		if !ok {
+			b.logger.Warn("unknown attributes type in timer event", "type", reflect.TypeOf(e.Attributes).String())
+			continue
+		}
+		b.logger.DebugContext(ctx, "scheduling timer to notify workflow worker")
+		time.AfterFunc(attr.At.Sub(time.Now()), func() { b.notifyWorkflowWorker(ctx) }) // TODO: cancel timer if the event gets cancelled
```

Comment:
Can we resolve that `TODO` in this PR? Might actually be fine not doing anything, the worker would be notified, but there wouldn't be a workflow task ready to execute so it should noop?

Reply:
That was my thought process as well - the worst thing that can happen is that the worker is notified when there's no task to execute. I left the `TODO` comment more as an opportunity to optimize this behavior, but I can also remove it.

On a related note - did you spot any cases when I forgot to trigger a notification, or are all cases covered?